### PR TITLE
Europa Lights

### PR DESCRIPTION
**Pull request by ghost**
Europa's lighting system, optimized and ported to /vg/station.

Stuff that's different from Europa:
- Wall lighting is much smoother and further extended than it is on Europa. This looks a lot better with our style and is less immersion breaking, plus you can actually see wallmounts.
- Every shadow has been reduced in size by 50% - 75% (a major optimization). They were normally as big as the entire light sprite and just tossed on as an overlay after being rotated, but now they're only a quarter or half of the original size and positioned correctly by a quadrants system.
- Shadows have a nice soft falloff on the sides as they get farther from the source, so they're less sharp and a bit prettier.
- Flickering lights. I could be wrong but I think Europa never actually implemented them, only left a define for them. I made them actually work and they're currently used for flames and sparks.

I might be forgetting something but I think that's all.

Current unresolved bugs (I was never really able to figure these out):
- ~~Directional lights (such as in flashlights) have incorrect pixel offsets until you change direction. It seems like the light might not being set to the correct direction for reasons unknown.~~ Fixed.
- Some areas may appear fully bright when in a certain position or direction, no solution was ever found to this even after clusterfack reported it on forums (http://www.byond.com/forum/?post=2262263)
- Space tiles can appear white as a blob/dead person, probably has something to do with changeturf code that adds a white overlay on the lighting plane to space tiles that are in an area with dynamic lighting.
- ~~Lights on light tiles are off center~~ Fixed.

Here's what it looks like.

https://a.pomf.cat/coqcmw.webm

:cl:
 * rscadd: Added Europa Lights lighting system.

### CODE REVIEW

**Comment by Shadowmech88** [comment]
>Directional lights (such as in flashlights) have incorrect pixel offsets until you change direction.

Have you tried calling whatever series of procs gets called when a light is rotated, automatically when the light is first turned on?

**Comment by ghost** [comment]
I believe I tried that already, but I'll have another go at it in a bit.
EDIT: still can't really figure it out.

**Comment by ghost** [comment]
Really makes me wish we could go back to wall-mounted objects not having black fucking gradients slapped on top.

**Comment by Chazach** [comment]
Everything looks so dark and unlit

**Comment by Exxion** [comment]
Oh those soft shadows look *so* much better than the hard ones did
Excellent job. This actually looks good now.

**Comment by Riekland** [comment]
This is way too fucking dark
How do the high efficiency lights look? How do floor lights look?
no really the lights need to be brighter how the fuck do you think this is okay

**Comment by Kammerjunk** [comment]
Are you sure light works like this?
It likes to bounce around and all

**Comment by Lokiamis** [comment]
I'm just here for the directional flashlights

**Comment by ghost** [comment]
I made light fixtures way brighter to compensate for 5 being the maximum range, and it actually worked pretty well.
Here's a video of a stroll through the station.
https://a.pomf.cat/jtgbik.webm
Obviously a few spots could use a little remapping, but overall it's fairly bright.

Light range is maxed at 5 because we don't have shadow sprites for any range higher than 5 and they were probably generated by a script that we don't have.

**Comment by yclatious** [comment]
Well, I hate to be a negative nancy but while the soft shadows do look much nicer, even the stronger brightness isnt enough, I gotta agree with riekland, this shit is way too dark, even with the much brighter altering.

Unless you amp it to "the station naturaly looks lighted without needing 15 lightbulbs per hallway", sorry, im not seeing it, the station isnt meant to look like a underwater isolated shithole bunker in the middle of the mariana trench, the optimization bonuses ive heard this will bring are great but if the lighting system wont work as a good one for natural gameplay, I cant really support it.

**Comment by ghost** [comment]
Well I can increase it, but someone has to create shadow sprites that look just like the current ones for light range 6 and 7, not to mention the amount of shadows will increase with each level of range so they'd have to figure out what the angle of those extra shadows would be.

I'll _try_ to ask Zuhayr if he has the extra sprites or a script for them, but beyond that there's not much I can do.

**Comment by Exxion** [comment]
Or you could just call it a mapping problem and add more lights
That's what Europa did, by my understanding
That would also allow you to get the shadows in nicer places

**Comment by Riekland** [comment]
It looks better but you need more light
if you have to add more lights add them
how do floor lights and high eff lights look
remember that you WILL need to remap every light on every station and away mission and asteroid before even thinking about merging this unless you can make them brighter fucko

**Comment by yclatious** [comment]
Again, hate to hate, but if its not practical for normal gameplay, if it looks like im staring at the abyss just because a clown with a gimmick removed a few lights or some greyshit busted the lights, I cant really say "sure, merge it" because it just will make playing a hassle, id need a flare to go by a hall without being in the dark if this passes.

I like the concept, I like the optimization, I like the style, but its just not really "playing" friendly.

**Comment by ghost** [comment]
@Riekland High efficiency lights are literally the same, they just consume less power. As for floor lights I hadn't even thought of those, they seem to be bugged at the moment so I'll work on a fix.
@yclatious I'm working on it, Zuhayr got me in touch with the guy who generated the shadows so hopefully once he's online I'll be able to get some sprites from him.

**Comment by CptWad** [comment]
/>oh look a downvote, bet its reikland
/>it is

**Comment by Riekland** [comment]
change is scary

**Comment by SangriaMonkey** [comment]
I don't like the idea of mood lighting on a space station.
Is this going to perform better than our current system?

**Comment by Exxion** [comment]
Yes.

**Comment by Shadowmech88** [comment]
It's about 50% more resource efficient, if I remember correctly.

**Comment by CptWad** [comment]
and a cheeki breeki iv dameeki to you too sir

**Comment by StarSmasher44** [comment]
I currently branch off of a bay build, but nonetheless the lighting system is mostly the same, so I was hoping to ask what exactly did you do to make it perform better? Perhaps I could take over some of these parts, we'd be grateful!

**Comment by ghost** [comment]
I love the look and atmosphere.  Better performance too?  Dog bless you.

File: code/modules/lighting/light_planes.dm [hunk 1]
```diff
@@ -0,0 +1,48 @@
+/obj/screen/plane
```
Comment (unid15) [inline code/modules/lighting/light_planes.dm]: /obj/abstract/screen/plane

File: code/modules/lighting/light_effect_cast.dm [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+		if(x_flip == 1 && y_flip == 1 && xy_swap == 0)
+			M.Translate(0, shadowoffset)
+		else if(x_flip == 1 && y_flip == 1 && xy_swap == 1)
+			M.Translate(shadowoffset / 2, shadowoffset / 2)
+		else if(x_flip == 1 && y_flip == -1 && xy_swap == 1)
+			M.Translate(-shadowoffset / 2, shadowoffset / 2)
+
+	//apply the transform matrix
+	I.transform = M
+	I.layer = 2
```
Comment (unid15) [inline code/modules/lighting/light_effect_cast.dm]: Defines for readability here

File: code/game/objects/items/weapons/power_cells.dm [hunk 3]
```diff
@@ -125,7 +125,7 @@
 				if(!user.light_range)
 					user.set_light(2,2,"#ffff00")
 					spawn(power_to_use/100 SECONDS)
-						user.set_light(0)
+					user.kill_light()
```
Comment (unid15) [inline code/game/objects/items/weapons/power_cells.dm]: indentation after spawn

File: code/modules/lighting/light_planes.dm [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+/obj/screen/plane/master
+	icon = 'icons/mob/screen1.dmi'
+	appearance_flags = NO_CLIENT_COLOR | PLANE_MASTER | RESET_TRANSFORM | RESET_COLOR | RESET_ALPHA
+	//list(null,null,null,"#0000","#000F")
+	color = LIGHTING_PLANEMASTER_COLOR  // Completely black.
+	plane = LIGHTING_PLANE_MASTER
+	mouse_opacity = 0
+
+//poor inheritance shitcode
+/obj/backdrop
```
Comment (unid15) [inline code/modules/lighting/light_planes.dm]: make this derived from screen objects too

**Comment by Zennyization** [comment]
@unusualcrow Any chance that you could make it so that as power fails in an area the wall-mounted lights begin to flicker? That'd be pretty cool.

**Comment by FudgePucker** [comment]
>50% less resource use

:+1:

**Comment by ghost** [comment]
@StarSmasher44 
> Every shadow has been reduced in size by 50% - 75% (a major optimization). They were normally as big as the entire light sprite and just tossed on as an overlay after being rotated, but now they're only a quarter or half of the original size and positioned correctly by a quadrants system.

@Zennyization I'll see if I can do something like that

**Comment by ghost** [comment]
[spooky](https://a.pomf.cat/ftmdyc.webm)

**Comment by Riekland** [comment]
I really like that flickering
If you can make the stations not dark and fix the overlapping looking like aids this will be very good

**Comment by ghost** [comment]
@Riekland what do you mean by overlapping looking like aids, what overlapping?

**Comment by Riekland** [comment]
my eyes

this

**Comment by ghost** [comment]
if you're referring to this specifically
meme
it's an issue with the way view in BYOND works i *think*, dunno if there's really any way around it

**Comment by StarSmasher44** [comment]
Could you perhaps tell me a bit about how that was done or show me the major part of code changed to get said result? I doubt a port would work on my end but I could always try to get the optimization part if possible.

**Comment by ghost** [comment]
@StarSmasher44 Well to begin, you export every shadow sprite and separate them. Most of them will only be within one quarter, or quadrant, of the actual image, since normally they're just thrown on top of the light (which is the same total size) with no regard for how big the shadow actually is.
untitled-1
So since you only need that one quadrant they're in, you can crop them into that quadrant of the image (has to be exact or else things will get fucky).

Some shadows, however, will be within two quadrants, so you'll only cut them in half.
untitled-1

Then you let those shadows get rotated as normal, and translate them into their correct positions based on how they're rotated, since rotation will tell you where they should be (if a quarter shadow is rotated to face the top left, it needs to be in the top left quadrant of the light).
So this code I added goes through every rotation possibility and translates the shadow overlay to its correct position.

	if(num == 1) //this means it's a quarter shadow
		if((x_flip == 1 && y_flip == 1 && xy_swap == 0) || (x_flip == -1 && y_flip == 1 && xy_swap == 1))
			M.Translate(shadowoffset, shadowoffset)
		else if((x_flip == 1 && y_flip == -1 && xy_swap == 0) || (x_flip == 1 && y_flip == 1 && xy_swap == 1))
			M.Translate(shadowoffset, 0)
		else if((xy_swap == 0 && x_flip == -y_flip) || (xy_swap == 1 && x_flip == -1 && y_flip == -1))
			M.Translate(0, shadowoffset)
	else //it's a half shadow
		if(x_flip == 1 && y_flip == 1 && xy_swap == 0)
			M.Translate(0, shadowoffset)
		else if(x_flip == 1 && y_flip == 1 && xy_swap == 1)
			M.Translate(shadowoffset / 2, shadowoffset / 2)
		else if(x_flip == 1 && y_flip == -1 && xy_swap == 1)
			M.Translate(-shadowoffset / 2, shadowoffset / 2)

That's all there is to it. Hope that helps.

**Comment by ghost** [comment]
also here's your answer to range 6 and 7
wellshit

so unless somebody has a way to create the proper shadows for those ranges it's pretty much impossible
i'll do some map edits to make sure nothing is too dark when lights are all on

@Riekland as for that corner issue, i'm working on implementing an algorithm to replace BYOND's shit and fix that

**Comment by ghost** [comment]
Can something be done about windows near space not lighting up the area they are near?  It is pretty weird given that we get enough direct sunlight to power solar arrays.

**Comment by Kammerjunk** [comment]
The perspective in this game is all over the place but let's say one turf is one square metre.
The speed of light is 299792458 metres per second (in a vacuum) according to a 2-second unsourced Google search.
It takes about 16.5 nanoseconds for light to stop working.

**Comment by ghost** [comment]
@unusualcrow Mind porting this to /tg/ after this is merged

**Comment by ghost** [comment]
what that man said ^.  this looks way better than goon lighting.

**Comment by ghost** [comment]
@MoreRobustThanYou I might, depends on if I have the time and how this turns out.

**Comment by ghost** [comment]
Will this make vampires OP

**Comment by ghost** [comment]
i certainly hope not

**Comment by ghost** [comment]
@Riekland here's your floor lights

**Comment by optimumtact** [comment]
You need to soften and wrap better around one tile edges

**Comment by optimumtact** [comment]
or soften your shadow sprites so they don't look so goddamn triangular, put in some blur at the edges damn

**Comment by CptWad** [comment]
next you tg people will want bagelstation

**Comment by ghost** [comment]
They ARE softened at the edges, if they seem sharp where there's two lights it's probably because of the way they get stacked on top of each other

**Comment by optimumtact** [comment]
The moment you guys merge this we're gonna get spammed with requests to include it, so we may as well get our criticisms out the way ahead of time

**Comment by Zennyization** [comment]
It's a /tg/ invades /vg/ episode

**Comment by ihadtoregisterforthis** [comment]
Since there's shaper shadows, wouldn't this need adittional remapping?
Are all handheld or portable lights directional?

100% tested?

**Comment by Riekland** [comment]
if flashlights are one direction now, are other man-portable lights and shit like the floodlights, helmet lights, exosuit lights, hardsuit lights, etc all one direction now or do they still provide an all round glow

**Comment by ghost** [comment]
@Riekland Most things that are handheld and point in one direction as a real flashlight would be directional, though they have a bit of a glow around them as well.
@ihadtoregisterforthis Still working on optimization and stuff, as well as remapping to make sure the station is fully lit.

**Comment by Riekland** [comment]
Most? What are the exceptions?

**Comment by ghost** [comment]
https://dl.dropboxusercontent.com/s/d14eau80kdw8rkj/2017-08-22_05-58-24.webm?dl=0

Bruh, fix this.

**Comment by ghost** [comment]
https://dl.dropboxusercontent.com/s/kooeuxthx8v5g7w/2017-08-22_06-02-24.webm?dl=0

and delete this

**Comment by ghost** [comment]
solars

b r u h
p l e a s e

If you need a dog damn flashlight to see your solar arrays, i've got bad news for you.

**Comment by ghost** [comment]
1. will do
2. yes newscaster lights are dumb 
3. already did this, just hadn't committed yet 
this should probably be marked as WIP since I'm still working on some things 

@Riekland whatever I think should be the exception when I go through and finish setting light types

**Comment by ghost** [comment]
For the record, I really want to see this go through, so I will help you find all of your little cases to knock out before people have a chance to whine and moan on live.

**Comment by ghost** [comment]
https://dl.dropboxusercontent.com/s/ubheoar469j4hm6/2017-08-22_06-38-40.webm?dl=0

b r u h

Holodeck sims need a smol amount of work too.

uy6tbu6n7

I personally always thought the beach was prettiest at night :)

**Comment by ghost** [comment]
the areas that ordinarily have no lighting such as holodeck and solars literally just need a single var set, will commit it when I'm home

**Comment by ghost** [comment]
For now the to do list is as follows:
- fix grues (may or may not be possible)
- remap lighting in each station
- ~~fix some light fixture issues~~
- ~~set area dynamic lighting~~
- optimize amount of overlays to reduce client-side lag
- rework dark sight (?)

**Comment by FudgePucker** [comment]
Remapping lighting on every single map will be a bitch unless this gets total merge priority and other people help with the mapping.
Another problem that might occur is with more lighting, the more power is consumed. How much power does an engine need to produce (not that big of a deal but is somewhat of a concern when,), and how long can the SMES roundstart power last (how long engineers have to set up an engine. If there is a coordinated team then this isn't an issue, if there is like one dude setting up and the rest of engineering is nonexistent/fucked off into space, then the station will turn dark.).
Other than those minor concerns I really like these changes. I feel like the change to flashlights is great, right now they turn you into miniature sun, maintenance will actually be scary and fixing up maint with lighting will be actually worth the time.

**Comment by ghost** [comment]
I'm not going to remap every light on every station, I'm just adding them in here and there wherever it's too dark.

**Comment by yclatious** [comment]
So, just about everywhere then :^)

**Comment by Riekland** [comment]
>I'm not going to remap every light on every station, I'm just adding them in here and there wherever it's too dark.

yes you are
if you just add lights to make the dark areas not dark you'll have a stupid amount of overlap with your lights and cause excessive power drain and maybe lag
you will either fix the light being too dark or you will replace every light everywhere
this isn't negotiable

**Comment by Exxion** [comment]
Remapping even one object on every map in the same PR is barely reasonable. Moving every single light definitively isn't.
Also, you can power the entire station right now with an upgraded treadmill and three hivelords. Any of the actual engines would have no problem with even twice as many lights. I doubt you'd run into power issues even if you put a light on every tile, though I haven't done the math on that.

**Comment by ghost** [comment]
@Riekland The lights won't overlap if you place them correctly. Even along straight walls there can be spaces where there is a gap of darkness wide enough to fit one light but not overlap into the next light over. Power drain isn't really much of a concern and any station with even slightly competent engineers will set up an engine that can handle the load just fine anyway.
You've never contributed anything to this repo and you're ordering me to do a massive amount of shit as if I work for you alone, fuck off.

**Comment by optimumtact** [comment]
it would be a lot easier to just create the 7tile shadow lengths, it can't be that hard right?

**Comment by ghost** [comment]
@optimumtact I totally would if I could, but the current shadow sprites are very exact and were rendered from a 3d scene, then modified. In addition, you have to know the angle for every additional shadow sprite that will be involved with that range.
It would also continue to decrease clientside performance which is bad enough as is.

**Comment by Shadowmech88** [comment]
And the setup for rendering shadow sprites from that 3D scene no longer exists?

**Comment by ghost** [comment]
correct

**Comment by Kammerjunk** [comment]
>The lights won't overlap if you place them correctly.

What happens if I build lights for, say, the ghetto bar on Box (which is already really dark even with the four fixtures in there)?

**Comment by ghost** [comment]
Nothing, overlapping doesn't matter, station just looks better and is more efficient without any light overlapping

**Comment by Tacolizard** [comment]
well tg will likely do here the same thing we did with goonlights; port it over, and then tone it down a ton so retards stop squealing about it

**Comment by ghost** [comment]
I'll work on this probably this weekend, too busy recently

**Comment by Tacolizard** [comment]
why do these shadows need to be rendered from a 3d scene? aren't they just triangular black blocks?

**Comment by ghost** [comment]
@Tacolizard They don't have to but it's difficult to make them without it because of the exact angle, size, and anti aliasing. The more range on each light, the more angles you will have to cover as well.

**Comment by Shadowmech88** [comment]
What's the status on this?

**Comment by FudgePucker** [comment]
How do helmet lights work?
What about mesons?

I am kinda worried this will break mining, considering its dark everywhere and you really need to be able to see goliaths before they get close to you. One hit can get a lucky broken bone and it spirals out of control and you end up being brutaily ripped apart or bleed to death while you slowly try to make your way to medical attention.

**Comment by ghost** [comment]
@Shadowmech88 
I just haven't gotten around to this, it needs a fair bit of remapping and I'm also stumped on optimization to the point where I'm not sure it's possible. Areas with lots of light lag quite a bit due to the size and amount of overlays and it's enough to be a problem, especially on bad computers.

@FudgePucker 
Helmet lights will have a fairly large radius.
Mesons act about the same as they always have.
Mining should be the same as always.

**Comment by AlphaBuffalo** [comment]
I hope this didnt die

**Comment by ghost** [comment]
@AlphaBuffalo At this point it's pretty much dead because I don't think it can be further optimized and even remotely light-dense areas run like shit, even on an i5. I know a lot of people will bitch their hearts out because of that so it's not worth the effort doing the rest of the work to just leave it at that and have it closed.
So with that I'm closing this.

**Comment by StarSmasher44** [comment]
Instead, is there a way you could apply some of said optimizations to the original lighting system? Just wondering

**Comment by ghost** [comment]
No.

**Pull request closed without merging**